Add fast blur algorithm

torfmaster opened the pull request:
This PR is an alternative to https://github.com/image-rs/image/pull/2302 which adds the corresponding functionality in the `image` crate. It adds a fast approximation of Gaussian blur to this library.

It is an attempt to solve https://github.com/image-rs/image/issues/986. For the discussion discussion please refer to the PR in `image` until it is decided where the functionality will go.

ripytide commented:
Thanks for opening another PR in this repo, I feel it is much better placed here. I don't have as much knowledge about the specific blurring algorithm but I am happy to help with the code quality and maybe adding some benchmarks.

cospectrum commented:
Some beautification is needed; otherwise, ci will not pass

cospectrum commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+        for j in 0..width {
+            for channel in 0..channel_num {
+                let val = vals[channel] / (2.0 * r as f32 + 1.0);
+                let val = if val < min_value {
```
Replace entire if {} else if {} else {} with val.clamp(min_value, max_value)

torfmaster replied:
fixed

cospectrum commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+) -> Vec<P> {
+    let channel_size = width * height;
+
+    let mut out_samples: Vec<P> = vec![P::from(0).unwrap(); channel_size * channel_num];
```
No need for explicit Vec type

torfmaster replied:
fixed

cospectrum commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+    /// Test blur doesn't panick when passed 0.0
+    fn test_fast_blur_zero() {
+        let image = RgbaImage::new(50, 50);
+        let _ = super::fast_blur(&image, -1.0);
```
Add use `super::*` above

torfmaster replied:
fixed

cospectrum commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+use std::cmp::{max, min};
```
All min/max function calls can be avoided

torfmaster replied:
fixed

cospectrum commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+    };
+    let w_u = w_l + 2.0;
+
+    let m_ideal = (12.0 * sigma * sigma
```
Expression is too long, I'm sure there is a way to simplify it

torfmaster replied:
It`s II (5) in https://peterkovesi.com/papers/FastGaussianSmoothing.pdf. It results from inverting the equation of ideal sigma and the one coming from repeated box blurs. I don't know how to further simplify this polynomial as all of the terms are linearly independent (as polynomials).

cospectrum replied:
```rs
let m_ideal = 0.25 * (n as f32) * (w_l + 3.0) - 3.0 * sigma.powi(2) * (w_l + 1.0).recip();
```

torfmaster replied:
Thank you, now the formula fits into one line.

cospectrum commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+    use image::RgbaImage;
+
+    #[test]
+    /// Test blur doesn't panick when passed 0.0
```
It's -1.0

torfmaster replied:
fixed already in original PR

torfmaster commented:
@cospectrum thank you for your review comments. I am going to address them here and in my PR on the `image` crate. However, as long there is no decision whether the code will land in `image` or in `imageproc` I think it makes sense to continue the review in the PR for the image crate.

torfmaster commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+    channel_num * idx + channel
+}
+
+fn my_fast_horizontal_blur<P: Primitive>(
```
I renamed this to `horizontal_fast_blur_half` in the original PR.

torfmaster commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+        / (-4.0 * (w_l) - 4.0);
+    let m = f32::round(m_ideal) as usize;
+
+    let mut box_sizes: Vec<usize> = vec![];
```
I chose the latter. As n is usually between 3 and 6 (and in our case it is always 3) the effect is little, that's why I optimized this code for readability.

cospectrum replied:
Remove the explicit `Vec<usize>` type for readability

torfmaster replied:
It's already gone (in the PR to the `image` crate).

torfmaster commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+    channel: usize,
+    channel_num: usize,
+) -> P {
+    let x = min(width as isize - 1, max(0, x)) as usize;
```
fixed (again in original PR)

torfmaster commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+/// Kovesi, P.:  Fast Almost-Gaussian Filtering The Australian Pattern
+/// Recognition Society Conference: DICTA 2010. December 2010. Sydney.
+pub fn fast_blur<P: Pixel>(
+    image_buffer: &ImageBuffer<P, Vec<P::Subpixel>>,
```
- [ ] fix this once it is decided that 'fast_blur` will land in `imageproc`

torfmaster commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+pub fn fast_blur<P: Pixel>(
+    image_buffer: &ImageBuffer<P, Vec<P::Subpixel>>,
+    sigma: f32,
+) -> ImageBuffer<P, Vec<P::Subpixel>> {
```
- [ ] fix this once it is decided that 'fast_blur` will land in `imageproc`

torfmaster commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+    out_samples
+}
+
+fn extended_f<P: Primitive>(
```
I created the function to increase the readability: it tells you that this function is the projection from R^2 to the bounded rectangle concatenated with f itself. Sure inlining would be shorter but the mathematical meaning would be forgotten.

torfmaster commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+    sigma: f32,
+) -> ImageBuffer<P, Vec<P::Subpixel>> {
+    let (width, height) = image_buffer.dimensions();
+    let mut samples = image_buffer.as_flat_samples().samples.to_vec();
```
I actually considered using the `Image` type plus the `put_pixel` method to implement the algorithm. However, I found that using slices is much more expressive (and presumably completely equivalent). If there is a point in using the `Image` type I can give it a try again.

torfmaster commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+        for (channel, value) in vals.iter_mut().enumerate().take(channel_num) {
+            *value = ((-(r as isize))..(r + 1) as isize)
+                .map(|x| {
+                    extended_f(samples, width, height, x, i as isize, channel, channel_num)
```
I experimented a bit with the number types involved and the result you see here is my (local) minimum of conversions. If you have an idea how to get rid of most of the conversions I am happy to implement this.

torfmaster commented on src/filter/fast_blur.rs:
```diff
@@ -0,0 +1,163 @@
+                let val = P::from(val).unwrap();
+
+                out_samples[channel_idx(channel, i + j * height, channel_num)] = val;
+                vals[channel] = vals[channel]
```
I extracted the index into a speaking variable. I hope this makes the code more readable.

torfmaster commented:
I just wanted to give a short status report (I also cross post this comment to https://github.com/image-rs/imageproc/pull/683)

 - I added a functional test showing (for an example) that the effect of `fast_blur` is close to the Gaussian blur implementation in this library
 - I added a benchmark that shows that fast blur is much faster than the current blur implementation
 - I think I addressed all code review comments

What would be the next steps?

cospectrum commented:
merged into `image`

torfmaster commented:
As the code is now in the `image` crate I will close this PR.
